Added CKEditor as current RTE and conversion function

bansalnitish opened the pull request:
This PR sets the flag to true which means we have CKEditor in as of now. It also adds a conversion function which will automatically convert all the html strings into the CKEditor format.

There were lots of tests associated with the front end. Since we swtiched to CKEditor a lot of them required updation. I have updated all and they should pass except the `editorFeatures`. I don't know why there are line breaks in the html when downloaded through history tag at the end of this line
`html: <p>enter 6 to continue</p>` 
![screenshot from 2018-07-09 04-00-23](https://user-images.githubusercontent.com/22434689/42425230-e28b7b4e-8337-11e8-83de-9cf3bae63384.png)

bansalnitish committed:
Added migration fucntion and tests for CKEditor

bansalnitish committed:
Set CKEditor as current RTE

AllanYangZhou commented:
Looking into it, I think the cause of that particular test issue is simply because CKEditor formats its html source output. Unlike textangular, which throws all its output html into one string on one line, CKEditor actually "prettifies" its source content (splits content up onto multiple lines and indents things):
![image](https://user-images.githubusercontent.com/8732147/42429250-d3bebe5a-82ec-11e8-88fe-beaa6547fdd3.png)

Although you *could* configure CKEditor to put all content on one line, I think its better to just modify this test case to match what CKEditor outputs.

seanlip commented:
I have no objections to doing that. I assume the HTML after the update by CKEditor is still fully valid according to our checks, right?

AllanYangZhou commented:
Yup. This is only about the cosmetics (not the structure) of the html--the html parser (e.g., BeautifulSoup during validation, or the RTE itself when editing) should be indifferent to whether the content is prettified or on one line.

seanlip commented:
OK, cool, thanks! Sounds like it should be an easy fix, then.

Thanks for taking a look!

AllanYangZhou reviewed (commented):
Most of the changes look fine--but you need to replace textAngular with ckEditor for suggestions too. Notice that the suggestion modal does not use `schema_based_html_editor`, so switching the flag does not enable CKEditor there. (I should have caught this in the earlier PR).

To do so now should be fairly simple, just modify `learner_view_suggestion_modal.html`. Notice that in textAngular for suggestions, the image component is disabled. Make sure the same happens for ckEditor, you just need to check for `componentDefns.requiresFs`.

bansalnitish committed:
Add CKEditor as RTE in suggestions

bansalnitish committed:
Merge remote-tracking branch 'upstream/develop' into ckeditor-migrate

bansalnitish committed:
Fixed issue with history testing

bansalnitish committed:
Fixed issue with TextAngularValidationAndMigration test

bansalnitish committed:
Fix issue with suggestion rendering

bansalnitish committed:
Fix linter issues

bansalnitish commented:
Hi @AllanYangZhou @seanlip 

I fixed the issue with history testing -- The text for the history was as follows:
```
1
      classifier_model_id: null
      2
      content:
      3
        content_id: content
      4
        html: '<p>this is card 2</p>
      5
       6
          '
      7
      content_ids_to_audio_translations:
...
```
Since while extracting lines, even indexes are treated as line number and odd indexes as text for that line, line 6 was being considered as text for line 5 which has no text actually. I added a check in `forms.js` to ensure proper extraction.

I also added CKEditor in the suggestion modal but I am facing an issue there:
We have this `ng-model="suggestionHtml"` in the `learner_view_suggestion_modal_directive.html`, which is a two way binding i.e any change to ng-model will do change is suggestionHtml and vice versa. But I printed the value of ng-model and suggestionHtml they turned out to be different. ng-model is the text we write in editor (it's updated perfectly fine) while the suggestionHtml always remain equal to original html. I'm not sure why this is the case.

There were 3 tests in `editorFeatures` 2 have passed this suggestion is the third one. 

EDIT : Also disabled image component in suggestions.

Thanks!

bansalnitish committed:
Disabled image in suggestions

AllanYangZhou commented on core/templates/dev/head/components/forms/FormBuilder.js:
```diff
@@ -521,6 +523,12 @@ oppia.directive('ckEditorRte', [
           ck.setData(wrapComponents(ngModel.$viewValue));
+
+          if (!canUseFs) {
+            $('.cke_button__oppiaimage').attr('onclick', '').unbind('click');
```
Rather than specifically disabling the image component, you should check whether `componentDefn.requiresFs` is true to determine whether a component requires using the filesystem (if it does, and `canUseFs` is false, then you disable. Please refer to the way textAngular handles this.

Also, please check if there is a cleaner way to disable a CKEditor toolbar button--manually adjusting the CSS does not seem correct. A quick search led me [here](https://stackoverflow.com/a/23521552), maybe that is helpful

bansalnitish replied:
Fixed this, PTAL!

bansalnitish commented:
Adding further to the above comment (just to make it more clear):
The error can be reproduced by visiting the local/test server, pick any exploration from the library and click on suggest changes. The submit button to submit the suggestion is always inactive.

The submit button is disabled by this line:
https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/exploration_player/learner_view_suggestion_modal_directive.html#L14

But since the value of ng-model has changed so should suggestionHtml.

AllanYangZhou reviewed (commented):
Left comments on the image-disabling implementation.

Re the submit button issue: Nitish and I discussed over DM, the issue will be addressed as described here: https://stackoverflow.com/a/22768720.

bansalnitish committed:
Added dot in ng-model to ensure the travis passes

bansalnitish committed:
Added comments

bansalnitish committed:
Improved image disabling

bansalnitish committed:
Made comment more clear

AllanYangZhou commented on core/templates/dev/head/pages/exploration_player/LearnerLocalNav.js:
```diff
@@ -47,7 +47,8 @@ oppia.controller('LearnerLocalNav', [
             $scope.isCurrentRTECKEditor = IS_CURRENT_RTE_CKEDITOR;
             $scope.description = '';
-            $scope.suggestionHtml = $scope.originalHtml;
+            $scope.suggest = {};
```
Nit: rename this to `suggestionData`.

bansalnitish replied:
Done!

AllanYangZhou commented on core/templates/dev/head/pages/exploration_player/learner_view_suggestion_modal_directive.html:
```diff
@@ -3,7 +3,7 @@ <h3>Suggest a Change</h3>
 <div class="modal-body protractor-test-exploration-suggestion-modal">
-  <ck-editor-rte ng-if="isCurrentRTECKEditor" ng-show="showEditor" ng-model="suggestionHtml"></ck-editor-rte>
+  <ck-editor-rte ng-if="isCurrentRTECKEditor" ng-show="showEditor" ng-model="suggest.suggestionHtml"></ck-editor-rte>
   <text-angular-rte ng-if="!isCurrentRTECKEditor" ng-show="showEditor" html-content="suggestionHtml"></text-angular-rte>
```
You should update `html-content` for textAngular too.

bansalnitish replied:
Removed the TextAngular stuff, PTAL!

AllanYangZhou commented on core/templates/dev/head/pages/exploration_player/LearnerLocalNav.js:
```diff
@@ -38,15 +38,22 @@ oppia.controller('LearnerLocalNav', [
+            $scope.isCurrentRTECKEditor = IS_CURRENT_RTE_CKEDITOR;
             $scope.description = '';
-            $scope.suggestionHtml = $scope.originalHtml;
+            // If there is a ng-model, you have to have a dot in there.
```
Instead of "you have to have a dot" better to explain something like "ng-model needs to bind to a property *of an object* on the scope (the property cannot sit directly on the scope)". Also, drop the last sentence that starts with "Due to this..."

Minor: you can use the shortlink: https://stackoverflow.com/q/12618342

bansalnitish replied:
Done!

AllanYangZhou commented on core/templates/dev/head/pages/exploration_player/learner_view_suggestion_modal_directive.html:
```diff
@@ -3,15 +3,17 @@ <h3>Suggest a Change</h3>
 
 <div class="modal-body protractor-test-exploration-suggestion-modal">
-  <text-angular-rte ng-show="showEditor" html-content="suggestionHtml"></text-angular-rte>
+  <!-- If there is a ng-model, you have to have a dot in there. -->
```
See above, "have a dot in there" is not very clear.

bansalnitish replied:
Done!

AllanYangZhou commented on core/tests/protractor_utils/forms.js:
```diff
@@ -631,8 +637,29 @@ var CodeMirrorChecker = function(elem) {
       var textArray = text.split('\n');
       for (var i = 0; i < textArray.length; i += 2) {
-        var lineNumber = textArray[i];
+        // CKEditor pretifies html and adds new lines.
```
Nit: prettifies

bansalnitish replied:
Done!

AllanYangZhou commented on core/tests/protractor_utils/forms.js:
```diff
@@ -637,15 +637,24 @@ var CodeMirrorChecker = function(elem) {
         var lineText = textArray[i + 1];
         var copy = lineText;
         var lineTextWithoutStartSpaces = copy.replace(/^\s+/g, '');
+        // To check if the line only contains a single digit.
```
Nit: `Checks if...`

bansalnitish replied:
Done!

seanlip commented on core/templates/dev/head/pages/exploration_player/LearnerLocalNav.js:
```diff
@@ -38,15 +38,22 @@ oppia.controller('LearnerLocalNav', [
+            // -model-does-not-update-controller-value/22768720#22768720.
+            // Due to this reason, suggestionHtml has been changed to
+            // suggest.suggestionHtml.
+            $scope.suggest = {};
```
Have a better name for this -- suggestionData perhaps? ("suggest" is an action, so not appropriate as a noun here). 

Also set the object's keys directly on-declaration, otherwise it becomes harder to tell what its contents are at any given time. Don't add new keys beyond the ones initially set. (This is why we typically use domain objects -- to avoid this sort of thing.)

bansalnitish replied:
Done!

seanlip commented on core/templates/dev/head/pages/exploration_player/learner_view_suggestion_modal_directive.html:
```diff
@@ -3,15 +3,17 @@ <h3>Suggest a Change</h3>
-  <text-angular-rte ng-show="showEditor" html-content="suggestionHtml"></text-angular-rte>
+  <!-- If there is a ng-model, you have to have a dot in there. -->
+  <ck-editor-rte ng-if="isCurrentRTECKEditor" ng-show="showEditor" ng-model="suggest.suggestionHtml"></ck-editor-rte>
+  <text-angular-rte ng-if="!isCurrentRTECKEditor" ng-show="showEditor" html-content="suggestionHtml"></text-angular-rte>
```
Why have both suggest.suggestionHtml and suggestionHtml? Shouldn't these two RTEs refer to the same thing?

Also, I think you should get rid of the isCurrentRTECKEditor flag and just remove the textAngular stuff. Once this PR is in, there's no going back, so there's no point having the switch.

bansalnitish replied:
Done!

bansalnitish commented:
HI @seanlip @AllanYangZhou 
All these tests are passing locally. I'm not sure why they are failing here.

bansalnitish committed:
Updated comments, variable and removed TextAngular stuff

bansalnitish committed:
Removed insertpre download

bansalnitish committed:
Fixed travis, moved ckeditor its js file, removed sanitizeHTMLForRTE, updated manifest file to maintain the format of the third party

apb7 reviewed (approved):
LGTM for commit [1a734d](https://github.com/oppia/oppia/pull/5229/commits/1a734dacf34cb9ddac2ccc27819f1198b2cd6ede) (Took a general pass and discussed with @bansalnitish about the changes done). Thanks!

bansalnitish commented:
Ah finally, tests have passed! :)

AllanYangZhou commented on core/templates/dev/head/components/CKEditorDirective.js:
```diff
@@ -0,0 +1,219 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
```
Rename this file to `CkEditorRteDirective.js`

bansalnitish replied:
done!

AllanYangZhou reviewed (approved):
After the one issue I noted is addressed, this looks good to me.

Note I am at work where I do not have oppia set up so I can't run these latest changes locally to inspect them visually. I trust that you've tested the rte yourself, both in the editor view with state content and interactions and in suggestions view.

bansalnitish committed:
Changed file name

seanlip commented on core/templates/dev/head/components/CkEditorRteDirective.js:
```diff
@@ -0,0 +1,219 @@
+        CKEDITOR.plugins.addExternal(
+          'sharedspace',
+          '/third_party/static/ckeditor-sharedspace-4.9.2/', 'plugin.js');
+        CKEDITOR.plugins.addExternal(
```
I noticed you'd deleted the "pre" plugin from some places. Do you still need this? Ditto below under extraPlugins.

If so please add a comment to explain why the path here is different from the others. If not please remove pre from all places.

bansalnitish replied:
Yeah, I need this. Done!

seanlip commented:
See inline comment. Once that's addressed feel free to merge once travis checks pass. Thanks!

bansalnitish committed:
Added comment to explain why the path of pre is different

seanlip commented on core/templates/dev/head/components/CkEditorRteDirective.js:
```diff
@@ -0,0 +1,221 @@
+        CKEDITOR.plugins.addExternal(
+          'sharedspace',
+          '/third_party/static/ckeditor-sharedspace-4.9.2/', 'plugin.js');
+        // Pre plugin is not available for 4.9.2 version of CKEditor. This is
```
Interesting -- you don't need to make a follow-up commit to address this, but I was wondering if you knew _why_ the pre plugin is not available? If they deprecated it for some reason then we probably want to know what that reason is so that we don't fall into the same problem that they (presumably) encountered...

bansalnitish replied:
No, I don't know. But yes I agree we should figure it out.

seanlip commented:
Ok. Can you find out? Check their mailing lists, github, etc

On Tuesday, July 10, 2018, Nitish Bansal <notifications@github.com> wrote:

> *@bansalnitish* commented on this pull request.
> ------------------------------
>
> In core/templates/dev/head/components/CkEditorRteDirective.js
> <https://github.com/oppia/oppia/pull/5229#discussion_r201475020>:
>
> > +        });
> +        buttonNames.pop();
> +        // All icons on the toolbar except the Rich Text components.
> +        var allIcons = ['undo', 'redo', 'bold', 'Italic', 'numberedList',
> +          'bulletedList', 'pre', 'indent', 'outdent'];
> +
> +        // Add external plugins.
> +        CKEDITOR.plugins.addExternal(
> +          'widget', '/third_party/static/ckeditor-widget-4.9.2/', 'plugin.js');
> +        CKEDITOR.plugins.addExternal(
> +          'lineutils',
> +          '/third_party/static/ckeditor-lineutils-4.9.2/', 'plugin.js');
> +        CKEDITOR.plugins.addExternal(
> +          'sharedspace',
> +          '/third_party/static/ckeditor-sharedspace-4.9.2/', 'plugin.js');
> +        // Pre plugin is not available for 4.9.2 version of CKEditor. This is
>
> No, I don't know. But yes I agree we should figure it out.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/5229#discussion_r201475020>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeyoJTodd594-pnAe_Eep51_SJFSmvks5uFQhNgaJpZM4VG7EC>
> .
>

bansalnitish commented:
Ya definitely. I'll figure it out.
